### PR TITLE
DRY sampler improvements

### PR DESCRIPTION
**Pull request by belladoreai**
I was asked by @p-e-w to split some of the changes from #6047 into a separate PR here.

This PR contains the data type performance improvement for DRY, and a minor fix to prevent crash on large repetitive inputs.

Edit: now also contains change to cap the max match length to 50

See the main PR for more info.

## Checklist:

- [x] I have read the [Contributing guidelines](https://github.com/oobabooga/text-generation-webui/wiki/Contributing-guidelines).

### CODE REVIEW

File: modules/sampler_hijack.py
Comment (p-e-w) [inline modules/sampler_hijack.py:239]: Please use descriptive identifiers like the rest of the function rather than single letters.

File: modules/sampler_hijack.py
Comment (p-e-w) [inline modules/sampler_hijack.py:251]: Here's an idea: What about instead of converting the tensor to a Python list, you convert it to a `np.array` (NumPy is already a dependency)?

This would give you access to `np.argwhere` etc., which might be quite a bit faster than a Python loop, while *also* being simpler and cleaner. And I doubt that NumPy has the same indexing overhead that Torch seems to have.

File: modules/sampler_hijack.py
Comment (p-e-w) [inline modules/sampler_hijack.py:268]: The cap can be **way** lower than 1000. Even 50 should be more than enough. The default `dry_base` is 1.75, and 1.75^50 > 10^12, which is a penalty that makes the token probability equal to zero for all practical purposes.

At which point I guess the Z algorithm becomes unnecessary, because the function is no longer quadratic even in the worst case, since the inner loop is bounded by a constant.

File: modules/sampler_hijack.py
Comment (belladoreai) [reply modules/sampler_hijack.py:251]: It seems that there are many different ways to iterate over elements in a NumPy array, and all of them are slower than iterating elements on a Python list: https://github.com/numpy/numpy/issues/16985

File: modules/sampler_hijack.py
Comment (belladoreai) [reply modules/sampler_hijack.py]: Correct

Comment (belladoreai) [reply modules/sampler_hijack.py]: I'd say let's merge this PR and close the other one?

File: modules/sampler_hijack.py
Comment (p-e-w) [reply modules/sampler_hijack.py:251]: Thanks for the info. Guess I have to revisit my simplistic assumption "if you want speed, use NumPy". Pretty crazy that this hasn't been optimized yet.

File: modules/sampler_hijack.py
Comment (p-e-w) [reply modules/sampler_hijack.py]: Yes, that's the most pragmatic solution IMO. I'd file the Z algorithm under "it's cool that it's possible, but the juice isn't quite worth the squeeze in this case".

File: modules/sampler_hijack.py
Comment (p-e-w) [inline modules/sampler_hijack.py:241]: This comment doesn't make sense anymore since the list elements are already raw integers.

File: modules/sampler_hijack.py
Comment (p-e-w) [inline modules/sampler_hijack.py:237]: :fire: empty line

**Review by p-e-w — APPROVED** [review]
LGTM modulo style nits.

**Comment by belladoreai** [comment]
@oobabooga Ready for merge

**Review by p-e-w — APPROVED** [review]
Looks good now! Probably makes sense to merge this before DRY is merged into master.

**Comment by jojje** [comment]
LGTM.  

Seems @p-e-w is also good with it. What's your view @oobabooga ?

It is an improvement over the original. _If_ further optimization turns out to be necessary, such can be done in separate PRs.

**Comment by Vhallo** [comment]
Good to see the performance issues being solved. Seems like it might be worthwhile to integrate this into Exllamav2 / TabbyAPI now as well?

**Comment by oobabooga** [comment]
Thanks for the reviews, merging now before merging DRY to the main branch.

**Pull request merged**
